File: windows/src/Elevate.Core/Models/AdminPortal.cs
namespace Elevate.Core.Models;

/// <summary>
/// A Microsoft admin portal that can be opened straight into one tenant, for the tenant menu's
/// "Open…" submenu. The Ibiza shells (Azure, Entra, Intune) take the tenant id as the first path
/// segment; the Microsoft 365 security shells (Defender, Purview) take it as <c>?tid=</c>. The
/// Microsoft 365 admin center has no tenant-in-URL form and is deliberately absent. None of these
/// carry the signed-in account: the browser session decides, and Microsoft's sign-in page prompts
/// with the tenant already fixed when it has none.
/// </summary>
public sealed record AdminPortal(string Title, string Host, bool TenantInQuery)
{
    public static readonly AdminPortal Azure = new("Azure Portal", "portal.azure.com", TenantInQuery: false);
    public static readonly AdminPortal Entra = new("Entra admin center", "entra.microsoft.com", TenantInQuery: false);
    public static readonly AdminPortal Intune = new("Intune admin center", "intune.microsoft.com", TenantInQuery: false);
    public static readonly AdminPortal Defender = new("Defender Portal", "security.microsoft.com", TenantInQuery: true);
    public static readonly AdminPortal Purview = new("Purview Portal", "purview.microsoft.com", TenantInQuery: true);

    /// <summary>Menu order.</summary>
    public static IReadOnlyList<AdminPortal> All { get; } = [Azure, Entra, Intune, Defender, Purview];

    /// <summary>The portal opened in <paramref name="tenantId"/>.</summary>
    public Uri Uri(string tenantId)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(tenantId);
        var escaped = System.Uri.EscapeDataString(tenantId);
        return new Uri(TenantInQuery ? $"https://{Host}/?tid={escaped}" : $"https://{Host}/{escaped}");
    }
}

File: macos/Sources/ElevateCore/Models/AdminPortal.swift
import Foundation

/// A Microsoft admin portal that can be opened straight into one tenant, for the tenant menu's
/// "Open…" submenu. The Ibiza shells (Azure, Entra, Intune) take the tenant id as the first path
/// segment; the Microsoft 365 security shells (Defender, Purview) take it as `?tid=`. The
/// Microsoft 365 admin center has no tenant-in-URL form and is deliberately absent. None of
/// these carry the signed-in account: the browser session decides, and Microsoft's sign-in page
/// prompts with the tenant already fixed when it has none.
public enum AdminPortal: String, CaseIterable, Sendable {
    case azure, entra, intune, defender, purview

    /// Menu title, in the order `allCases` lists them.
    public var title: String {
        switch self {
        case .azure: "Azure Portal"
        case .entra: "Entra admin center"
        case .intune: "Intune admin center"
        case .defender: "Defender Portal"
        case .purview: "Purview Portal"
        }
    }

    /// The portal opened in `tenantId`.
    public func url(tenantId: String) -> URL {
        let tenant = Self.escape(tenantId)
        switch self {
        case .azure, .entra, .intune: return URL(string: "https://\(host)/\(tenant)")!
        case .defender, .purview: return URL(string: "https://\(host)/?tid=\(tenant)")!
        }
    }

    private var host: String {
        switch self {
        case .azure: "portal.azure.com"
        case .entra: "entra.microsoft.com"
        case .intune: "intune.microsoft.com"
        case .defender: "security.microsoft.com"
        case .purview: "purview.microsoft.com"
        }
    }

    /// Percent-encodes everything but RFC 3986 unreserved characters, so the id is safe as a
    /// path segment and as a query value alike. Tenant ids are GUIDs, so this is belt and braces.
    private static func escape(_ s: String) -> String {
        let unreserved = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
        return s.addingPercentEncoding(withAllowedCharacters: unreserved) ?? s
    }
}
